### DISCUSSION TITLE
how to schedule cron tasks in directus V9 ?

### DISCUSSION THREAD

**Discussion by StartMeUp** [directus/directus#4120] [Archive]
Hi there

I've noticed that the whole app engine is actually stored and working in node_modules. 

How would you implement scheduled jobs / tasks in the root directory ? I don't want to touch the core directus files since an update might erase my cutom code. 

Normally I would have required node-cron in an index.js or server.js, but app.js/server.js/start.js are located in node_modules and I don't want to mess with them.

Any suggestion ?

**Comment by rijkvanzanten** [comment]
There's a feature request to add a native way to schedule tasks (see https://github.com/directus/directus/discussions/3062). Until that's a thing, you could use [the `init` hook](https://docs.directus.io/guides/api-hooks/#event-format-options) to setup a task using `node-cron` like you suggested

**Reply by StartMeUp** [reply]
Many thanks, I'll look into it.

**Reply by juancarlosjr97** [reply]
Yes, using `init` works very good :)

Nice idea @StartMeUp

**Reply by tspvivek** [reply]
@juancarlosjr97 Could you please help by sharing the code. I am trying to add cron to update a item after every 1 hour.

**Reply by juancarlosjr97** [reply]
Sure @tspvivek!

I use a docker configuration.
Dockerfile
```
FROM ....

AllVariablesHere

COPY custom-entrypoint.sh .
RUN chmod +x ./custom-entrypoint.sh
CMD ["command", "./custom-entrypoint.sh"]

EXPOSE 80
```

custom-entrypoint.sh
```
#!/usr/bin/env bash

npx directus bootstrap

npm install --save node-cron@3.0.0
npm install --save domo-sdk@2.0.1

exec npx directus start
```

Then you have this hook on init

```
module.exports = () => {
    return {
      init: async () => {
        const cron = require("node-cron");

        cron.schedule('* * * * *', () => {
          console.log('running a task every minute');
        });
      },
    };
  };
```

And then you are done :) 

If you have any other question, feel free to drop any questions.

**Reply by tspvivek** [reply]
@juancarlosjr97 Thanks it worked well. If possible could you please let me know how to access services (ItemsService) inside init event function.

**Reply by juancarlosjr97** [reply]
Sure @tspvivek 

You only need to add to the init hook the following

```
module.exports = ({ services }) => {
  return {
    init: async () => {
      const { ItemsService } = services;
.....
    },
  };
};
```

If you have any other question, just let me know. Also, if you are not part of the Discord community, join us https://discord.com/invite/directus, many people online and great support available :)

**Reply by tspvivek** [reply]
@juancarlosjr97 Thanks, you saved my day. I am already part of directus discord community, will post further queries there :)

**Reply by rijkvanzanten** [reply]
@juancarlosjr97 @tspvivek I'm planning on natively supporting the cron syntax as a "regular" hook, so you can do something like:

```js
module.exports = () => {
  return {
    '* * * * *': async () => {
      console.log('running a task every minute');
    },
  };
};
```

**Reply by tspvivek** [reply]
@rijkvanzanten This would be much simpler and clearer approach.

**Reply by rijkvanzanten** [reply]
@tspvivek You got lucky, managed to add it into rc.74 🙂 

Starting with rc.74, you can set it up as follows:

```js
module.exports = () => {
  return {
    'cron(* * * * *)': async () => {
      console.log('running a task every minute');
    },
  };
};
```

**Reply by tspvivek** [reply]
@rijkvanzanten So quick, this one is great addon to this futuristic product. Gonna update my app to rc.74 now. :+1:

**Reply by juancarlosjr97** [reply]
@rijkvanzanten awesome work! Love it!
